### DISCUSSION TITLE
Coroutines and related abstractions

### DISCUSSION THREAD

**Discussion by pvgoran** [zio/zio#7199] [Ideas]
When trying to implement a task combinator that repeatedly executes the task and reports the execution status to the provided `ZSink` (https://github.com/zio/zio/discussions/7195), it occured to me that the cleanest way to achieve this would involve the coroutime abstraction. Sadly, this abstraction apparently doesn't exist in ZIO. Here I'm trying to explore what it could look like.

I may be wrong, but I see a coroutine as a computation that, when executed, (1) can produce one or more intermediate outputs and (2) eventually finishes with a result. Intermediate outputs are given to the caller of the coroutine, which processes them and then returns control to the coroutine (or not). While the caller processes an intermediate output, the coroutine's execution is suspended, and is resumed when the caller says so. As an optional feature, we can also envision the possibility to pass intermediate input from the caller back to the coroutine when the latter is resumed by the former.

So, we could have the type `ZCoroutine[R, E, +TOutput, -TResume, +TResult]`, which describes a coroutine that runs in the `R` environment, can fail with a error `E`, emits intermediate output values of type `TOutput`, accepts back intermediate input values of type `TResume`, and finally produces the end result of type `TResult`.

From the outside, a coroutine looks similar to `ZStream`, except that a stream doesn't produce the end result and doesn't have the possibility to receive intermediate input values. However, the internal structure of coroutine would be different: it would be a computation like a `ZIO`, with the ability to yield back to the caller at any point inside the computation (from a deeply nested sub-computation within loops, conditionals, internal functions and so on).

A coroutine can be (naively) modelled like this:

```scala
  type ZFunction[-R, +E, -TParam, +TResult] = TParam => ZIO[R, E, TResult]
  type ZCoroutine[R, E, +TOutput, -TResume, +TResult] = ZFunction[R, E, TOutput, TResume] => ZIO[R, E, TResult]
```

So it's just a function that receives a "yield function" of type `TParam => ZIO[R, E, TResult]` and produces a computation.

For example:

```scala
  val sampleCoroutine : ZCoroutine[Any, Throwable, String, Int, Int] = { yieldFunction =>
    Ref.make(0).flatMap { counter =>
      def increaseCounter(incr : Int) : Task[Unit] =
        counter.update(_ + incr)

      zio.Console.printLine("Hello coroutine")
      *> yieldFunction("Start").flatMap(input => increaseCounter(input))
      *> ZIO.whenZIO(zio.Random.nextDouble.map(_ > 0.5)) {
        yieldFunction("Condition satisfied") // intermediate input ignored here
      }
      *> ZIO.foreach(1 to 5) { i =>
        if i % 2 == 0
          then yieldFunction(s"Value $i").flatMap(input => increaseCounter(input))
          else ZIO.unit
      }
      *> counter.get
    }
  }
```

Now, we need a way to run a coroutine. `ZStream` has a `ZSink` that complements it; similarly, we might want a complimentary abstraction for a coroutine, something that accepts coroutine's intermediate inputs, provides intermediate outputs in return, has an optional itnernal state that could be influenced by intermediate outputs and influences intermediate inputs, and produces the final result from the coroutine's final result and the accumulated internal state. I don't readily have a name for this abstraction, it could be, for example, `ZCoprocessor[R, E, -TOutput, +TResume, TCoResult, +TResult]`:

```scala
  trait ZCoprocessorInstance[R, E, -TOutput, +TResume, TCoResult, +TResult]
  {
    def yieldFunction : ZFunction[R, E, TOutput, TResume]
    def resultFunction : ZFunction[R, E, TCoResult, TResult]
  }

  type ZCoprocessor[R, E, -TOutput, +TResume, TCoResult, +TResult] = ZIO[R, E, ZCoprocessorInstance[R, E, TOutput, TResume, TCoResult, TResult]]
```

`ZCoprocessor` would be created using factory functions and various transformations, much like `ZSink` is.

With this in mind, we could have the "run" function:

```scala
  def runCoroutine[R, E, TOutput, TResume, TCoResult, TResult](
    coroutine : ZCoroutine[R, E, TOutput, TResume, TCoResult],
    coprocessor : ZCoprocessor[R, E, TOutput, TResume, TCoResult, TResult]
  ) : ZIO[R, E, TResult] =
  {
    ...
  }
```

This function would create a co-processor instance, instantiate the coroutine computation, handle its intermediate outputs (feeding them sequentially to the co-processor instance, even if the coroutine tries to call `yieldFunction` concurrently), wait for completion of coroutine computation, deactivate the `yieldFunction` after completion (so that, if possible forked computations within the coroutine call the `yieldFunction`, it would fail), and map the end result via the co-processor instance.

This model doesn't include the possibility to **not** returning control to the co-routine after it produces an itnermediate output; this possibility could be implemented by using a special `Shutdown` error that the co-procesor's `yieldFunction` is allowed to return:

```scala
def yieldFunction : ZFunction[R, E | Shutdown, TOutput, TResume]
```

(Not sure how it would actually interact with the coroutine computation: return the `Shutdown` error to it from `yieldFunction`? Interrupt it? This needs more thought.)

Now, all of this is more like a dream rather than an actual proposal. It's probably quite difficult to even write a good API for this, let alone implement it. I may be able to create a rudimentary, non-production-ready implementation for my own use, though. :) And if anyone would actually create a proper implementation, it could be a noteworthy addition to the library.

**Comment by iravid** [comment]
Have you looked at ZChannel? It is exactly
your ZCoroutine and is the underlying abstraction for streams and sinks.

**Reply by pvgoran** [reply]
> Have you looked at ZChannel?

I tried to, but I have troubles understanding its semantics.

> It is exactly your ZCoroutine

I saw some similarity, but `ZChannel`'s properties, and the way they are created, didn't look appropriate to me.

Specifically,

1. Are `ZChannel`'s input and output synchronized (reads happen at the same time as writes)?
2. How can I produce a `ZChannel` from, say, the `sampleCoroutine` function defined in the opening post?

**Comment by adamgfraser** [comment]
Yes, there are actually several data types like this in ZIO.

One is `Schedule`, which is something you can feed inputs into and for each input it will produce an output and say whether it is willing to accept further inputs or is done. You can use a `Driver` to step through a schedule feeding inputs to it and deciding whether to call new inputs based on the outputs.

Another is `ZChannel`, which can accept multiple inputs as well as eventually a terminal value and produce multiple outputs and eventually a terminal value. A variety of stream operators such as running a stream with a sink, transducing a stream with a sink, or transforming a stream with a pipeline all drive the channel by feeding inputs to it one at a time.

There is no guarantee that a channel will emit an output for each input, and in fact some channels may emit multiple outputs with no input or consume multiple inputs with no output, but you can create channels that emit an output for each input.

You can create channels with a variety of channel constructors . For example, you can use `ZChannel.readWith` to describe a channel that reads some input and `ZChannel.fail`, `Channel.write` or `ZChannel.succeed` to describe whether the channel should fail, emit an output, or complete with a terminal value based on that input. By recursing and maintaining internal state you can describe much more complex logic.

**Reply by pvgoran** [reply]
`Schedule` is quite close both to the "co-processor" abstraction that I envisioned and the `yieldFunction` thing that a coroutine works with. However, a `Schedule` is supposed to be infallible (no errors possible), and it doesn't deal with the final result. Both of these properties restrict its viability as a model for "co-processor". They don't get in the way of using a `Schedule.Driver` as a `yieldFunction`; however, this kind of usage would need the co-routine computation to surround the calls to the `Schedule.Driver` with the boiler-plate code.

`ZChannel` is a way too broad abstraction, it has too many variations of behaviour and use. Using it would require too much attention to maintain the appropriate discipline and to avoid inappropriate manipulations. Besides, I expect that the `ZChannel` representation of a computation (with multiple "yield" points deep within the computation structure) would be unnecessarily complex and obscure. And I want to describe my logic in a way that doesn't involve additional complexity and unnatural representation. :)

That being said, a `ZChannel` can probably be used as an implementation detail for the coroutine mechanics.
